### PR TITLE
feat: add socket channel support to in-app message client

### PR DESCRIPTION
**Pull request by thomaswhyyou**
This PR adds real time updates to in-app message client, via our switchboard socket API (a corresponding PR [here](https://github.com/knocklabs/switchboard/pull/2963)).

Here's an overview of how things are structured:
```
1 x Knock client (with 1 x socket conn) 
  -> 1 x In-app channel client (with 1 x in-app socket driver)
    -> N x In-app message client(s) (joins or leaves via In-app channel client)
```
* The core knock client is instantiated (per API key and a user id), and this is where a socket connection lives. 
* In-app channel client is instantiated (per in-app channel) wrapping the knock client and instantiating an in-app "socket driver" with the socket connection from the knock client. 
  * In-app socket driver is the one that manages 1) subscribing/unsubscribing to socket channels 2) receiving socket events and setting to its internal reactive (tanstack) store 3) notifying any subscribed in-app message clients when a relevant socket event is arrived.
* In-app message clients are instantiated per message type and combinations of filter params. There can be one or more in-app message clients that are talking to our in-app messages API via in-app channel client. 

TLDR: we handle multiple instances of in-app message clients receiving real time updates from our socket API through a single socket connection, which is managed by an in-app socket driver.

### CODE REVIEW

File: packages/client/src/clients/in-app-messages/message-client.ts [hunk 1]
```diff
@@ -14,10 +15,14 @@ import {
 
 /**
  * Manages realtime connection to in app messages service.
+ *
+ * TODO: Rename to InAppMessagesClient to singular?
```
Comment (connorlindsey) [inline packages/client/src/clients/in-app-messages/message-client.ts:19]: This was feedback from Chris to match the other clients like `UsersClient` and `Preferences` although there are several that are singular. No strong opinion either way

**Comment by thomaswhyyou** [comment]
* **#230** <a href="https://app.graphite.dev/github/pr/knocklabs/javascript/230?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `next`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @thomaswhyyou and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/client/package.json [hunk 2]
```diff
@@ -77,6 +77,7 @@
     "axios-retry": "^3.1.9",
     "eventemitter2": "^6.4.5",
     "jwt-decode": "^4.0.0",
+    "nanoid": "^3.0.0",
```
Comment (thomaswhyyou) [inline packages/client/package.json:80]: it's pretty tiny: https://bundlephobia.com/package/nanoid@3.0.0

File: packages/client/src/clients/in-app-messages/channel-client.ts [hunk 3]
```diff
@@ -16,19 +18,30 @@ import {
 export class InAppChannelClient {
   public store: InAppStore;
 
+  private socketDriver: InAppMessageSocketDriver | undefined;
```
Comment (thomaswhyyou) [inline packages/client/src/clients/in-app-messages/channel-client.ts:21]: Open to a different name if anyone has better suggestions.

File: packages/client/src/clients/in-app-messages/socket-driver.ts [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+export class InAppMessageSocketDriver {
+  // Mapping of live channels by topic. Note, there can be one or more in-app
+  // message client(s) that can subscribe.
+  private channels: Record<ChannelTopic, Channel>;
+
+  // Mapping of query params for each in-app message client, partitioned by its
+  // reference id, and grouped by channel topic. It's a double nested object
+  // that looks like:
+  //  {
+  //    "in_app:card:...": {
```
Comment (thomaswhyyou) [inline packages/client/src/clients/in-app-messages/socket-driver.ts:52]: The inner object is what gets sent to the socket API as params: 
https://github.com/knocklabs/switchboard/pull/2963/files#diff-aa5f2245deb4ebc08de405e1f9cac816d0e942bcfffbfcd1693e8d3594d24427R30

File: packages/client/src/clients/in-app-messages/socket-driver.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+      channelForTopic.leave();
+      delete channels[topic];
+    }
+
+    this.params = partitionedParams;
+    this.channels = channels;
+  }
+
+  private setInbox(payload: WithAttn<SocketEventPayload>) {
+    const { attn, ...rest } = payload;
```
Comment (thomaswhyyou) [inline packages/client/src/clients/in-app-messages/socket-driver.ts:176]: `attn` field is provided by the socket API to indicate which in-app message clients should be notified of the incoming socket event based on respective query params. 
https://github.com/knocklabs/switchboard/pull/2963/files#diff-aa5f2245deb4ebc08de405e1f9cac816d0e942bcfffbfcd1693e8d3594d24427R137

File: packages/client/src/clients/in-app-messages/channel-client.ts (hunk 3 above)
Comment (cjbell) [reply packages/client/src/clients/in-app-messages/channel-client.ts:21]: I think `InAppMessageSocketManager` is better than `Driver`

File: packages/client/src/clients/in-app-messages/message-client.ts [hunk 6]
```diff
@@ -233,6 +238,30 @@ export class InAppMessagesClient {
...
+  unsubscribe() {
+    this.channelClient.unsubscribe(this);
+  }
+
+  // This is a callback function that will be invoked when a new socket event
+  // relevant for this message client is received (and if subscribed).
+  async handleSocketEvent(payload: SocketEventPayload) {
+    switch (payload.event) {
+      case SocketEventType.MessageCreated:
+        return await this.fetch();
```
Comment (cjbell) [inline packages/client/src/clients/in-app-messages/message-client.ts:254]: Per my point in the Switchboard PR, no need to refetch here now!

File: packages/react-core/src/modules/in-app-messages/hooks/useInAppMessages.ts [hunk 7]
```diff
@@ -48,14 +48,15 @@ export const useInAppMessages = <
   );
 
   useEffect(() => {
-    const intervalId = setInterval(() => {
-      inAppMessagesClient.fetch();
-    }, 2500);
+    inAppMessagesClient.fetch();
+    inAppMessagesClient.subscribe();
```
Comment (cjbell) [inline packages/react-core/src/modules/in-app-messages/hooks/useInAppMessages.ts:52]: Extreme nit but I would probably suggest flipping these

**Review by cjbell — APPROVED** [review]
this looks awesome

File: packages/client/src/clients/in-app-messages/channel-client.ts [hunk 8]
```diff
@@ -89,4 +102,22 @@ export class InAppMessagesChannelClient {
       },
     }));
   }
+
+  /*
+   * Socket
+   */
+
+  subscribe(client: InAppMessagesClient) {
```
Comment (connorlindsey) [inline packages/client/src/clients/in-app-messages/channel-client.ts:110]: May be worth adding some logs (using `knock`) for potential debugging

**Review by connorlindsey — APPROVED** [review]
Looks awesome, the comments are super helpful for understanding the socket implementation

File: packages/client/src/clients/in-app-messages/channel-client.ts (hunk 3 above)
Comment (thomaswhyyou) [reply packages/client/src/clients/in-app-messages/channel-client.ts:21]: Yep that works too, will update.

File: packages/client/src/clients/in-app-messages/channel-client.ts (hunk 8 above)
Comment (thomaswhyyou) [reply packages/client/src/clients/in-app-messages/channel-client.ts:110]: Yep, will update!

File: packages/client/src/clients/in-app-messages/message-client.ts (hunk 6 above)
Comment (thomaswhyyou) [reply packages/client/src/clients/in-app-messages/message-client.ts:254]: Replied in the switchboard PR thread - I did consider doing something smarter here with the in-app message included in the event payload instead of refetching but I realized it wasn't going to be as simple as I initially thought. 

For reference, more details in this thread: https://github.com/knocklabs/switchboard/pull/2963#discussion_r1803877703

**Review by cjbell — APPROVED** [review]
🚢 🚢🚢
